File: src/engine/taxonomy.ts
// ARGUS-P v2 taxonomy — faithful TS port of argus_p/taxonomy.py
//
// The central correction in v2 is that a "person" is not one thing: a founder,
// a KOL, an investor, an agency, and a community member are evaluated against
// different evidence and different failure modes.

export enum SubjectClass {
  FOUNDER = "FOUNDER",
  KOL = "KOL",
  INVESTOR = "INVESTOR",
  ADVISOR = "ADVISOR",
  AGENCY = "AGENCY",
  MEMBER = "MEMBER",
}

// Classes are NOT mutually exclusive. A single subject can hold several roles
// at once (the common case: a builder who also invests and advises). Each held
// role is scored on its own track and the composite verdict is governed by the
// most severe role. Roles are never averaged together.

// Pseudonymity is NOT a flag. Risk lives in behaviour and outcomes (rugs,
// dumping into own promotion, paid-cabal-then-collapse, fabrication,
// impersonation), not in identity state. There is NO hard identity gate on
// pseudonymity. Only genuine impersonation / identity fraud blocks publication.
export const HARD_IDENTITY_GATE_CLASSES: Set<SubjectClass> = new Set();

// Reward for verifiable real-world identity, added to each role's score.
export const DOX_BONUS: Record<string, number> = { Confirmed: 5, Probable: 3 };

// --------------------------------------------------------------------------
// Founder / team controlled vocabularies
// --------------------------------------------------------------------------

export enum VentureOutcome {
  ACTIVE = "Active",
  PAUSED = "Paused",
  IPO = "IPO",
  ACQUISITION = "Acquisition",
  ACQUIHIRE = "Acquihire",
  ORDERLY_WINDDOWN = "OrderlyWindDown",
  FAILURE = "Failure",
  SILENT_SHUTDOWN = "SilentShutdown",
  RUG = "Rug",
  EXPLOIT = "Exploit",
  UNKNOWN = "Unknown",
}

// A public pause is NOT a silent shutdown: SilentShutdown requires evidence of
// abandonment, not merely the absence of a launch the analyst happened to find.
export const NON_TERMINAL = new Set<VentureOutcome>([
  VentureOutcome.ACTIVE,
  VentureOutcome.PAUSED,
  VentureOutcome.UNKNOWN,
]);

export const POSITIVE_OUTCOMES = new Set<VentureOutcome>([
  VentureOutcome.IPO,
  VentureOutcome.ACQUISITION,
]);
export const SEVERE_OUTCOMES = new Set<VentureOutcome>([VentureOutcome.RUG]);
export const NEGATIVE_OUTCOMES = new Set<VentureOutcome>([
  VentureOutcome.SILENT_SHUTDOWN,
  VentureOutcome.RUG,
]);

export enum FounderPattern {
  SERIAL_SUCCESS = "SerialSuccess",
  PROVEN_ONCE = "ProvenOnce",
  MIXED = "Mixed",
  SERIAL_FAILURE = "SerialFailure",
  RUG_HISTORY = "RugHistory",
  UNPROVEN = "Unproven",
  FIRST_VENTURE = "FirstVenture",
}

export function classifyFounderPattern(
  outcomes: (VentureOutcome | string)[],
): FounderPattern {
  const outs = outcomes.map((o) => o as VentureOutcome);
  const completed = outs.filter((o) => !NON_TERMINAL.has(o));

  if (outs.some((o) => SEVERE_OUTCOMES.has(o))) return FounderPattern.RUG_HISTORY;
  if (outs.length === 0) return FounderPattern.FIRST_VENTURE;
  if (completed.length === 0) return FounderPattern.UNPROVEN;

  const successes = completed.filter((o) => POSITIVE_OUTCOMES.has(o)).length;
  const failures = completed.filter(
    (o) => o === VentureOutcome.SILENT_SHUTDOWN || o === VentureOutcome.FAILURE,
  ).length;

  if (successes >= 2 && failures === 0) return FounderPattern.SERIAL_SUCCESS;
  if (successes === 1 && failures === 0) return FounderPattern.PROVEN_ONCE;
  if (successes === 0 && failures >= 2) return FounderPattern.SERIAL_FAILURE;
  if (successes >= 1 && failures >= 1) return FounderPattern.MIXED;
  return FounderPattern.UNPROVEN;
}

export interface VentureRecord {
  outcome?: VentureOutcome | string;
  investors?: string[];
  acquirer?: string | null;
  current_backers?: string[];
}

export interface RepeatBackingResult {
  repeat_backers: string[];
  from_successful_exit: boolean;
  strength: "none" | "weak" | "strong";
}

const norm = (name: string) => String(name).trim().toLowerCase();

export function repeatBackingSignal(ventures: VentureRecord[]): RepeatBackingResult {
  const current = new Set<string>();
  const repeat = new Set<string>();
  let fromSuccess = false;

  for (const v of ventures) {
    if (v.outcome === VentureOutcome.ACTIVE) {
      for (const b of v.current_backers ?? []) current.add(norm(b));
    }
  }
  for (const v of ventures) {
    const priorParties = new Set<string>((v.investors ?? []).map(norm));
    if (v.acquirer) priorParties.add(norm(v.acquirer));
    const overlap = [...priorParties].filter((p) => current.has(p));
    if (overlap.length) {
      overlap.forEach((o) => repeat.add(o));
      if (POSITIVE_OUTCOMES.has(v.outcome as VentureOutcome)) fromSuccess = true;
    }
  }
  let strength: "none" | "weak" | "strong";
  if (repeat.size === 0) strength = "none";
  else if (fromSuccess) strength = "strong";
  else strength = "weak";

  return {
    repeat_backers: [...repeat].sort(),
    from_successful_exit: fromSuccess,
    strength,
  };
}

// --------------------------------------------------------------------------
// Investor controlled vocabularies
// --------------------------------------------------------------------------

export enum FundTier {
  ANGEL = "Angel",
  SUPER_ANGEL = "SuperAngel",
  LAUNCHPAD = "Launchpad",
  MICRO_VC = "MicroVC",
  VC_TIER3 = "VC_Tier3",
  VC_TIER2 = "VC_Tier2",
  VC_TIER1 = "VC_Tier1",
  UNKNOWN = "Unknown",
}

export const TIER_RANK: Record<FundTier, number> = {
  [FundTier.ANGEL]: 1,
  [FundTier.SUPER_ANGEL]: 2,
  [FundTier.LAUNCHPAD]: 2,
  [FundTier.MICRO_VC]: 3,
  [FundTier.VC_TIER3]: 4,
  [FundTier.VC_TIER2]: 5,
  [FundTier.VC_TIER1]: 6,
  [FundTier.UNKNOWN]: 0,
};

export enum TestimonialVerdict {
  CORROBORATED = "Corroborated",
  PARTIAL = "PartiallyCorroborated",
  UNCONFIRMED = "Unconfirmed",
  CONTRADICTED = "Contradicted",
}

// --------------------------------------------------------------------------
// Shared evidence-status vocabulary (carried from v1, unchanged)
// --------------------------------------------------------------------------

export const IDENTITY_CONFIDENCE = [
  "Confirmed",
  "Probable",
  "Unverified",
  "SuspectedImpersonation",
] as const;
export type IdentityConfidence = (typeof IDENTITY_CONFIDENCE)[number];

export const VERIFICATION_STATUS = ["Verified", "Reported", "Rumor"] as const;
export const WALLET_LINK_TIERS = ["SelfDoxxed", "InvestigatorAttributed", "Inferred"] as const;
export const PROMO_DISCLOSURE = ["Disclosed", "Undisclosed", "Unknown"] as const;

File: engine_reference/argus_p/taxonomy.py
"""
ARGUS-P v2 taxonomy
===================

Defines the subject classes and the controlled vocabularies each class
depends on. The central correction in v2 is that a "person" is not one
thing: a founder, a KOL, an investor, an agency, and a community member
are evaluated against different evidence and different failure modes.

Nothing in this file changes the frozen entity-resolution contract
(see entity.py). It only adds classification on top of it.
"""

from enum import Enum


class SubjectClass(str, Enum):
    FOUNDER  = "FOUNDER"    # founders and core team / technical contributors
    KOL      = "KOL"        # influencers, callers, paid promoters
    INVESTOR = "INVESTOR"   # angels, super angels, launchpads, micro VCs, VCs
    AGENCY   = "AGENCY"     # marketing / growth / market-making contractors
    MEMBER   = "MEMBER"     # community, ambassadors, moderators
    ADVISOR  = "ADVISOR"    # named advisors / board / strategic to projects

# Classes are NOT mutually exclusive. A single subject can hold several roles
# at once (the common case: a builder who also invests and advises). Each held
# role is scored on its own track and the composite verdict is governed by the
# most severe role. Roles are never averaged together.


# Pseudonymity is NOT a flag. Many legitimate builders and investors in crypto
# operate under a persistent pseudonym with a verifiable on-chain or shipped
# track record (e.g. 0xdeveloper of $BNKR). Risk lives in behaviour and
# outcomes (rugs, dumping into own promotion, paid-cabal-then-collapse,
# fabrication, impersonation), not in identity state.
#
# Accordingly there is NO hard identity gate on pseudonymity. The only identity
# condition that blocks publication is genuine impersonation or identity fraud
# (SuspectedImpersonation), handled in the orchestrator. Instead of penalising
# pseudonymity, the model REWARDS disclosure: a verified real-world identity
# earns a doxxing bonus (DOX_BONUS) on top of the identity axis.
HARD_IDENTITY_GATE_CLASSES = set()   # intentionally empty; see note above

# Reward for verifiable real-world identity, added to each role's score.
DOX_BONUS = {"Confirmed": 5, "Probable": 3}


# --------------------------------------------------------------------------
# Founder / team controlled vocabularies
# --------------------------------------------------------------------------

class VentureOutcome(str, Enum):
    ACTIVE            = "Active"             # ongoing, no exit yet
    PAUSED            = "Paused"             # publicly paused / deprioritised, not abandoned (neutral)
    IPO               = "IPO"               # public listing (strong positive)
    ACQUISITION       = "Acquisition"        # genuine strategic/financial buy (positive)
    ACQUIHIRE         = "Acquihire"          # team absorbed, product sunset (ambiguous)
    ORDERLY_WINDDOWN  = "OrderlyWindDown"    # closed cleanly, capital returned (neutral)
    FAILURE           = "Failure"            # failed transparently (neutral / mild neg)
    SILENT_SHUTDOWN   = "SilentShutdown"     # disappeared quietly, no comms (negative)
    RUG               = "Rug"               # exit scam (severe negative, cap trigger)
    EXPLOIT           = "Exploit"            # hacked (context dependent)
    UNKNOWN           = "Unknown"

# Non-terminal outcomes are excluded from the founder-pattern completed set.
# A public pause is NOT a silent shutdown: SilentShutdown requires evidence of
# abandonment (no comms, no distribution, disappearance), not merely the
# absence of a launch the analyst happened to find.
NON_TERMINAL = {VentureOutcome.ACTIVE, VentureOutcome.PAUSED, VentureOutcome.UNKNOWN}

POSITIVE_OUTCOMES = {VentureOutcome.IPO, VentureOutcome.ACQUISITION}
SEVERE_OUTCOMES   = {VentureOutcome.RUG}
NEGATIVE_OUTCOMES = {VentureOutcome.SILENT_SHUTDOWN, VentureOutcome.RUG}


class FounderPattern(str, Enum):
    SERIAL_SUCCESS = "SerialSuccess"   # 2+ real exits, no rug
    PROVEN_ONCE    = "ProvenOnce"      # 1 real exit, no rug
    MIXED          = "Mixed"           # successes and failures
    SERIAL_FAILURE = "SerialFailure"   # 2+ silent shutdowns / failures, no success
    RUG_HISTORY    = "RugHistory"      # any confirmed rug as a principal (overrides)
    UNPROVEN       = "Unproven"        # no completed ventures
    FIRST_VENTURE  = "FirstVenture"


def classify_founder_pattern(outcomes):
    """Given an iterable of VentureOutcome, derive a founder-pattern label.

    A confirmed rug as a principal overrides everything else: it is the
    single most predictive negative signal for a repeat founder.
    """
    outs = [o if isinstance(o, VentureOutcome) else VentureOutcome(o) for o in outcomes]
    completed = [o for o in outs if o not in NON_TERMINAL]

    if any(o in SEVERE_OUTCOMES for o in outs):
        return FounderPattern.RUG_HISTORY
    if not outs:
        return FounderPattern.FIRST_VENTURE
    if not completed:
        return FounderPattern.UNPROVEN

    successes = sum(1 for o in completed if o in POSITIVE_OUTCOMES)
    failures  = sum(1 for o in completed if o in (VentureOutcome.SILENT_SHUTDOWN,
                                                  VentureOutcome.FAILURE))
    if successes >= 2 and failures == 0:
        return FounderPattern.SERIAL_SUCCESS
    if successes == 1 and failures == 0:
        return FounderPattern.PROVEN_ONCE
    if successes == 0 and failures >= 2:
        return FounderPattern.SERIAL_FAILURE
    if successes >= 1 and failures >= 1:
        return FounderPattern.MIXED
    return FounderPattern.UNPROVEN


def repeat_backing_signal(ventures):
    """Detect whether sophisticated prior backers re-backed the founder.

    `ventures` is a list of dicts with keys:
      outcome, investors (list), acquirer (str|None), current_backers (list)
    The strongest positive venture signal is a prior acquirer or prior
    investor in a *successful* venture choosing to back the new one again.

    Returns a dict: {repeat_backers: [...], from_successful_exit: bool,
                     strength: 'none'|'weak'|'strong'}.
    """
    current = set()
    repeat = set()
    from_success = False
    # current backers are recorded on the active/most-recent venture
    for v in ventures:
        if (v.get("outcome") in (VentureOutcome.ACTIVE, VentureOutcome.ACTIVE.value)):
            current |= set(map(_norm, v.get("current_backers", []) or []))
    for v in ventures:
        outcome = v.get("outcome")
        prior_parties = set(map(_norm, v.get("investors", []) or []))
        if v.get("acquirer"):
            prior_parties.add(_norm(v["acquirer"]))
        overlap = prior_parties & current
        if overlap:
            repeat |= overlap
            if outcome in POSITIVE_OUTCOMES or outcome in (VentureOutcome.IPO.value,
                                                           VentureOutcome.ACQUISITION.value):
                from_success = True
    if not repeat:
        strength = "none"
    elif from_success:
        strength = "strong"
    else:
        strength = "weak"
    return {"repeat_backers": sorted(repeat),
            "from_successful_exit": from_success,
            "strength": strength}


def _norm(name):
    return str(name).strip().lower()


# --------------------------------------------------------------------------
# Investor controlled vocabularies
# --------------------------------------------------------------------------

class FundTier(str, Enum):
    ANGEL        = "Angel"          # individual, small checks
    SUPER_ANGEL  = "SuperAngel"     # prolific individual
    LAUNCHPAD    = "Launchpad"      # IDO / launch platform (distinct model)
    MICRO_VC     = "MicroVC"        # institutional, sub ~50M typical
    VC_TIER3     = "VC_Tier3"
    VC_TIER2     = "VC_Tier2"
    VC_TIER1     = "VC_Tier1"       # established, brand-name
    UNKNOWN      = "Unknown"

TIER_RANK = {
    FundTier.ANGEL: 1, FundTier.SUPER_ANGEL: 2, FundTier.LAUNCHPAD: 2,
    FundTier.MICRO_VC: 3, FundTier.VC_TIER3: 4, FundTier.VC_TIER2: 5,
    FundTier.VC_TIER1: 6, FundTier.UNKNOWN: 0,
}


class TestimonialVerdict(str, Enum):
    CORROBORATED        = "Corroborated"          # public ack + claimed relationship confirmed
    PARTIAL             = "PartiallyCorroborated"  # some interaction, relationship unconfirmed
    UNCONFIRMED         = "Unconfirmed"            # no public trace either way
    CONTRADICTED        = "Contradicted"           # denial, distancing, or FUD present


# --------------------------------------------------------------------------
# Shared evidence-status vocabulary (carried from v1, unchanged)
# --------------------------------------------------------------------------

IDENTITY_CONFIDENCE = ("Confirmed", "Probable", "Unverified", "SuspectedImpersonation")
VERIFICATION_STATUS = ("Verified", "Reported", "Rumor")
WALLET_LINK_TIERS   = ("SelfDoxxed", "InvestigatorAttributed", "Inferred")
PROMO_DISCLOSURE    = ("Disclosed", "Undisclosed", "Unknown")
